File: Daily/2024/Feb/0368_Largest_Divisible_Subset.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size(), max = 0, index = -1;
        vector<int> count(n), pre(n), res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n; i++)
        {
            count[i] = 1, pre[i] = -1;
            for(int j = i - 1; j >= 0; j--)
            {
                if(nums[i] % nums[j] == 0 && count[j] + 1 > count[i])
                    count[i] = count[j] + 1, pre[i] = j;
            }
            if(count[i] > max)
                max = count[i], index = i;
        }
        while(index != -1)
        {
            res.push_back(nums[index]);
            index = pre[index];
        }
        return res;
    }
};

File: Daily/2025/Apr/0368_Largest_Divisible_Subset.go
package main
import (
	"fmt"
	"sort"
)

func largestDivisibleSubset(nums []int) []int {
	// initialize
	dp := make([]int, len(nums))
	prev := make([]int, len(nums))
	res := []int{}
	max_val, index := 1, 0
	for i := 0; i < len(dp); i++ {
		dp[i] = 1
		prev[i] = -1
	}
	sort.Ints(nums)
	// build dp
	for i := 0; i < len(nums); i++ {
		for j := i - 1; j >= 0; j-- {
			if nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i] {
				dp[i] = dp[j] + 1
				prev[i] = j
				if dp[i] > max_val {
					max_val = dp[i]
					index = i
					break
				}
			}
		}
	}
	// build ans
	for index != -1 {
		res = append(res, nums[index])
		index = prev[index]
	}
    
	return res
}

func main() {
	arr1 := []int{1,2,3,4,5,6,7,8,9,10,11,12,13,36}
	arr2 := []int{4,8,10,240}

	fmt.Println(largestDivisibleSubset(arr1))
	fmt.Println(largestDivisibleSubset(arr2))
}